### ISSUE TITLE
Add support for running "Deploy" GitHub workflow when GitHub pre-release/release is created

### ISSUE THREAD

**Issue by angela-tran** [cal-itp/eligibility-server#286]
- [x] Update `deploy.yml` to check for pre-release/release events
- [x] Add step(s) to `deploy.yml` to run appropriate checks (run tests, etc.) if pre-release/release
- [x] Add logic to name the Docker tag appropriately when we see a GitHub pre-release/release
- [x] Create Docker tag with Git tag name too
- [x] Configure [tag protections](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-tag-protection-rules) so that only benefits-admin team can push tags that trigger deployment (same sort of protection as what we have with branch protections for current deployment model)
- [x] Configure [Required reviewers](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#required-reviewers) on GitHub deployment environments for test/prod to provide an additional opportunity to check that the deployment looks correct

**Comment by angela-tran** [comment]
> - [ ] Configure [tag protections](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-tag-protection-rules) so that only benefits-admin team can push tags that trigger deployment (same sort of protection as what we have with branch protections for current deployment model)

While setting up tag protection, I noticed some UI for using repository rules instead:

![image](https://github.com/cal-itp/eligibility-server/assets/25497886/b0c604e4-8201-4e5c-9442-73e5323f3f91)

![image](https://github.com/cal-itp/eligibility-server/assets/25497886/67c07cd7-2889-44bd-8b5b-35b5039feb3c)

Repository rules had their GA release in [July 2023](https://github.blog/2023-07-24-github-repository-rules-are-now-generally-available/) and are more flexible than branch protection and tag protection.

I set up [two repository rules](https://github.com/cal-itp/eligibility-server/rules) to protect the creation/update/deletion of tags: [one rule for pre-release tags](https://github.com/cal-itp/eligibility-server/rules/232244) and [one for release tags](https://github.com/cal-itp/eligibility-server/rules/232271). Anyone with read access to the repository can view these rules to help determine why their action was blocked.

Someone who is not on the rule's bypass list will see an error message when trying to do a protected action:

![image](https://github.com/cal-itp/eligibility-server/assets/25497886/7a6ad1dc-3e5f-4852-a35c-8616dd5e3c2c)

Someone who is on the bypass list will be able to bypass the protection:

![image](https://github.com/cal-itp/eligibility-server/assets/25497886/d81bb945-6ddd-420e-b98e-90d4d7ca2a64)

**Comment by angela-tran** [comment]
Because repository rules show you what your target pattern matches, I noticed that the glob pattern we're using for release tags in `azure-pipelines.yml` is too permissive - `20??.??.?*` matches more than just release tags. 

For the release-tag repository rule, I used a simpler pattern that is more strict about the end of the value.

I will create a new issue to update our trigger pattern in [MST's pipeline](https://github.com/cal-itp/eligibility-server/blob/dev/terraform/mst/azure-pipelines.yml#L10) and [SBMTD's pipeline](https://github.com/cal-itp/eligibility-server/blob/dev/terraform/sbmtd/azure-pipelines.yml#L10).
